File: backend/erphub-api/microservices/retail-api/src/main/java/com/zaphirio/retailapi/operation/quote/service/QuoteService.java
package com.zaphirio.retailapi.operation.quote.service;

import com.zaphirio.retailapi.operation.quote.dto.*;
import java.time.LocalDate;
import java.util.List;
import java.util.UUID;

/**
 * Service interface for Quote operations.
 */
public interface QuoteService {

    /**
     * Create a new quote.
     *
     * @param request the create request
     * @param tenantId the tenant ID
     * @return the created quote response
     */
    QuoteResponse create(CreateQuoteRequest request, Long tenantId);

    /**
     * Get quote by ID.
     *
     * @param id the quote ID
     * @param tenantId the tenant ID
     * @return the quote response
     */
    QuoteResponse getById(UUID id, Long tenantId);

    /**
     * Get all quotes (paginated would be better for production).
     *
     * @param tenantId the tenant ID
     * @return list of quote responses
     */
    List<QuoteResponse> getAll(Long tenantId);

    /**
     * Get all quotes for a customer.
     *
     * @param customerId the customer ID
     * @param tenantId the tenant ID
     * @return list of quote responses
     */
    List<QuoteResponse> getByCustomerId(UUID customerId, Long tenantId);

    /**
     * Get all quotes for a customer in a specific branch.
     *
     * @param customerId the customer ID
     * @param branchId the branch ID
     * @param tenantId the tenant ID
     * @return list of quote responses
     */
    List<QuoteResponse> getByCustomerIdAndBranchId(UUID customerId, UUID branchId, Long tenantId);

    /**
     * Get quote by status.
     *
     * @param status the quote status
     * @param tenantId the tenant ID
     * @return list of quote responses
     */
    List<QuoteResponse> getByStatus(String status, Long tenantId);

    /**
     * Get active (valid) quotes.
     *
     * @param tenantId the tenant ID
     * @return list of quote responses
     */
    List<QuoteResponse> getActive(Long tenantId);

    /**
     * Get quotes by creation date range.
     *
     * @param startDate the start date
     * @param endDate the end date
     * @param tenantId the tenant ID
     * @return list of quote responses
     */
    List<QuoteResponse> getByDateRange(LocalDate startDate, LocalDate endDate, Long tenantId);

    /**
     * Get quotes expiring soon.
     *
     * @param daysUntilExpiration number of days to check
     * @param tenantId the tenant ID
     * @return list of quote responses
     */
    List<QuoteResponse> getExpiringQuotes(int daysUntilExpiration, Long tenantId);

    /**
     * Update a quote.
     *
     * @param id the quote ID
     * @param request the update request
     * @param tenantId the tenant ID
     * @return the updated quote response
     */
    QuoteResponse update(UUID id, UpdateQuoteRequest request, Long tenantId);

    /**
     * Send a quote to customer (change status to SENT).
     *
     * @param id the quote ID
     * @param tenantId the tenant ID
     * @return the updated quote response
     */
    QuoteResponse sendQuote(UUID id, Long tenantId);

    /**
     * Accept a quote (change status to ACCEPTED).
     *
     * @param id the quote ID
     * @param tenantId the tenant ID
     * @return the updated quote response
     */
    QuoteResponse acceptQuote(UUID id, Long tenantId);

    /**
     * Reject a quote (change status to REJECTED).
     *
     * @param id the quote ID
     * @param tenantId the tenant ID
     * @return the updated quote response
     */
    QuoteResponse rejectQuote(UUID id, Long tenantId);

    /**
     * Convert quote to sale.
     *
     * @param id the quote ID
     * @param saleId the sale ID to link
     * @param tenantId the tenant ID
     * @return the updated quote response
     */
    QuoteResponse convertToSale(UUID id, UUID saleId, Long tenantId);

    /**
     * Delete a quote.
     *
     * @param id the quote ID
     * @param tenantId the tenant ID
     */
    void delete(UUID id, Long tenantId);

    /**
     * Generate next quote number for a given year.
     *
     * @param year the year
     * @param tenantId the tenant ID
     * @return the next quote number (e.g., Q001/2026)
     */
    String generateNextQuoteNumber(int year, Long tenantId);
}

File: frontend/web-client/src/app/core/services/quote.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { environment } from '../../../environments/environment';
import { Observable } from 'rxjs';

export interface QuoteItem {
  id?: string;
  productId: string;
  description: string;
  quantity: number;
  unitPrice: number;
  discountPercentage: number;
  taxRate: string;
  lineTotal?: number;
  taxAmount?: number;
  lineTotalWithTax?: number;
}

export interface Quote {
  id?: string;
  customerId: string;
  branchId: string;
  locationId?: string;
  number: string;
  documentType: string;
  issueDate: string;
  expirationDate: string;
  totalPrice: number;
  status: string;
  saleId?: string;
  observations?: string;
  items: QuoteItem[];
}

export interface CreateQuoteRequest {
  customerId: string;
  branchId: string;
  locationId?: string;
  number: string;
  documentType: string;
  issueDate: string;
  expirationDate: string;
  totalPrice: number;
  observations?: string;
  items: QuoteItem[];
}

export interface UpdateQuoteRequest {
  expirationDate?: string;
  observations?: string;
}

/**
 * Service for Quote API operations.
 */
@Injectable({ providedIn: 'root' })
export class QuoteService {
  private readonly http = inject(HttpClient);
  private readonly apiUrl = `${environment.apiUrl}/api/quotes`;

  /**
   * Create a new quote.
   */
  create(request: CreateQuoteRequest, tenantId: string): Observable<Quote> {
    return this.http.post<Quote>(this.apiUrl, request, {
      headers: { 'X-Tenant-ID': tenantId }
    });
  }

  /**
   * Get quote by ID.
   */
  getById(id: string, tenantId: string): Observable<Quote> {
    return this.http.get<Quote>(`${this.apiUrl}/${id}`, {
      headers: { 'X-Tenant-ID': tenantId }
    });
  }

  /**
   * Get all quotes.
   */
  getAll(tenantId: string): Observable<Quote[]> {
    return this.http.get<Quote[]>(this.apiUrl, {
      headers: { 'X-Tenant-ID': tenantId }
    });
  }

  /**
   * Get quotes by customer.
   */
  getByCustomerId(customerId: string, tenantId: string): Observable<Quote[]> {
    return this.http.get<Quote[]>(`${this.apiUrl}/customer/${customerId}`, {
      headers: { 'X-Tenant-ID': tenantId }
    });
  }

  /**
   * Get quotes by status.
   */
  getByStatus(status: string, tenantId: string): Observable<Quote[]> {
    const params = new HttpParams().set('status', status);
    return this.http.get<Quote[]>(`${this.apiUrl}/status`, {
      params,
      headers: { 'X-Tenant-ID': tenantId }
    });
  }

  /**
   * Get active quotes (not expired, not converted).
   */
  getActive(tenantId: string): Observable<Quote[]> {
    return this.http.get<Quote[]>(`${this.apiUrl}/active`, {
      headers: { 'X-Tenant-ID': tenantId }
    });
  }

  /**
   * Get expired quotes.
   */
  getExpired(tenantId: string): Observable<Quote[]> {
    return this.http.get<Quote[]>(`${this.apiUrl}/expired`, {
      headers: { 'X-Tenant-ID': tenantId }
    });
  }

  /**
   * Get quotes by date range.
   */
  getByDateRange(startDate: string, endDate: string, tenantId: string): Observable<Quote[]> {
    const params = new HttpParams()
      .set('startDate', startDate)
      .set('endDate', endDate);
    return this.http.get<Quote[]>(`${this.apiUrl}/date-range`, {
      params,
      headers: { 'X-Tenant-ID': tenantId }
    });
  }

  /**
   * Update a quote.
   */
  update(id: string, request: UpdateQuoteRequest, tenantId: string): Observable<Quote> {
    return this.http.put<Quote>(`${this.apiUrl}/${id}`, request, {
      headers: { 'X-Tenant-ID': tenantId }
    });
  }

  /**
   * Convert quote to sale.
   */
  convertToSale(id: string, tenantId: string): Observable<Quote> {
    return this.http.post<Quote>(`${this.apiUrl}/${id}/convert-to-sale`, {}, {
      headers: { 'X-Tenant-ID': tenantId }
    });
  }

  /**
   * Reject a quote.
   */
  reject(id: string, tenantId: string): Observable<Quote> {
    return this.http.post<Quote>(`${this.apiUrl}/${id}/reject`, {}, {
      headers: { 'X-Tenant-ID': tenantId }
    });
  }

  /**
   * Delete a quote.
   */
  delete(id: string, tenantId: string): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`, {
      headers: { 'X-Tenant-ID': tenantId }
    });
  }
}
